flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2738,6 +2968,17 @@ class ArrayExprLowering {
   }
 
 private:
+  struct PushSemant {
```

Comment:
There is already a `Fortran::common::Restorer` and `Fortran::common::ScopeSet()` that I think does pretty much what you are doing here:

https://github.com/flang-compiler/f18-llvm-project/blob/cc23e9993b5a6ac16455171981289eff3c633ee3/flang/include/flang/Common/restorer.h#L23

Reply:
OK. I'll take a look at it.